Common/Arm64Emitter.cpp

```diff
@@ -3918,17 +4033,32 @@ void ARM64FloatEmitter::MOVI2F(ARM64Reg Rd, float value, ARM64Reg scratch, bool
 }
 
 // TODO: Quite a few values could be generated easily using the MOVI instruction and friends.
-void ARM64FloatEmitter::MOVI2FDUP(ARM64Reg Rd, float value, ARM64Reg scratch) {
+void ARM64FloatEmitter::MOVI2FDUP(ARM64Reg Rd, float value, ARM64Reg scratch, bool negate) {
+	_assert_msg_(!IsSingle(Rd), "%s doesn't support singles", __FUNCTION__);
 	// TODO: Make it work with more element sizes
-	// TODO: Optimize - there are shorter solution for many values
 	ARM64Reg s = (ARM64Reg)(S0 + DecodeReg(Rd));
 	int ival;
 	memcpy(&ival, &value, 4);
+	uint8_t imm8;
 	if (ival == 0) {  // Make sure to not catch negative zero here
-		EOR(Rd, Rd, Rd);
+		// Prefer MOVI 0, which may have no latency on some CPUs.
+		MOVI(32, Rd, 0);
+		if (negate)
+			FNEG(32, Rd, Rd);
+	} else if (negate && FPImm8FromFloat(-value, &imm8)) {
+		FMOV(32, Rd, imm8);
```

Comment:
Does this (and the other cases except the else) really do the DUP thing and replicate to all lanes? https://developer.arm.com/documentation/dui0801/h/A64-SIMD-Vector-Instructions/DUP--vector--general-?lang=en

Reply:
FMOV is a vector instruction directly.  It sets all the lanes.  This is from the ARM64 ARM:

```
if cmode<0> == '1' && op == '0' then
    imm32 = imm8<7>:NOT(imm8<6>):Replicate(imm8<6>,5):imm8<5:0>:Zeros(19);
    imm64 = Replicate(imm32, 2)
```

It later replicates this 64 bit constant to both 64-bit lanes if Q.

-[Unknown]

Reply:
Oh right, so I assume it's the same for FNEG. I guess I'm still in the ARM32 mindset where that would not work.